x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexing/AsyncTwoPhaseIndexer.java

```diff
@@ -315,6 +310,14 @@ private IndexerState finishAndSetState() {
                 throw new IllegalStateException("Indexer job encountered an illegal state [" + prev + "]");
             }
         });
+
+        if (callOnStop.get()) {
```

Comment:
Since this doesn't need the intermediate `stopping` state anymore, could we just check `updatedState` and avoid the atomic booleans?

I suppose it'd also trigger if was already stopped and that's not what you want?

Reply:
@polyfractal `stopping` still provides a call to `doSaveState` when `checkState` is called against it. 
But, changing these branches to
```
if(updatedState == IndexerState.STOPPED) {
   onStop();
else if(udpatedState == IndexerState.ABORTING){
   onAbort();
}
```
Maybe could work.

Reply:
@davidkyle what do you think? That would make it out of step with `onAbort` but, we could possibly remove at least that atomic boolean

Reply:
Eh, if it requires jumping through hoops it's probably not worth it.  Just was looking to see if there was an easy way to avoid the atomics.

Reply:
Carry on :+1: